### DISCUSSION TITLE
Github Actions Secrets passwords with special characters show undefined behavior

### DISCUSSION THREAD

**Discussion by askb** [community/community#61323] [Actions] [answered]
### Select Topic Area

Bug

### Body

Hello GH team:

I am currently working on a packer workflow file using GitHub Actions, and I am encountering a critical issue related to the handling of special characters, such as '$', in the credentials stored within GitHub secrets. The problem arises when these credentials are read in the workflow file, as they truncate all characters following the '$'. This issue has caused significant debugging efforts, consuming a substantial amount of time.

Here is an example to illustrate the problem:

```
Sample secret set on GitHub Actions: `^!@#$%^&*pas$word123!^%!#`
Reads from the workflow file "${{ secrets.MYSEC }}": returns `^!@#$%^&*pas!^%!#`
```

To provide more context, I have implemented the proper quoting rules in accordance with GitHub's documentation, as indicated here: [GitHub Actions - Encrypted secrets](https://docs.github.com/en/actions/security-guides/encrypted-secrets). However, despite adhering to these guidelines, the outcome remains undefined, leading to the unexpected behavior when referencing the secrets containing special characters.

The workflow file in question can be accessed through this link: [Packer Workflow File](https://github.com/askb/releng-builder/blob/master/.github/workflows/packer-verify.yaml)

Upon running the workflow, I encountered the following error message:

```
Error on running the workflow:
This character is not used within the language.
Error: Unsupported operator

|    6: test_pass = '\''{{ secrets.TESTPASS }}'\''
| 
| Single quotes are not valid. Use double quotes (") to enclose strings.

```

Considering the significance of this issue, I would like to inquire if this is a known problem with GitHub Actions. Furthermore, I am curious whether there are other special characters that should be avoided in passwords when working with GitHub secrets. It is essential to address this matter, as limiting passwords to exclude certain charsets might adversely impact the overall entropy.

Lastly, I would greatly appreciate it if you could provide any documentation or workarounds available to resolve this issue. As an additional piece of information, I want to highlight that running the same test locally on my laptop or another node using Jenkins successfully validates the templates without encountering any issues.

Thank you for your attention to this matter. I look forward to your prompt response and assistance in resolving this matter.

**Comment by ChristopherHX** [comment]
Your linked workflow does some unsafe string interpolation.
Some run step content will be interpreted by 3 languages
- GitHub Actions (${{ env.varname }})
- bash ($varnams)
- yaml

The docs you have mentioned pass secrets via `env` and not via `run`, to solve bash script injection. While yaml script injection is still possible, due to writing a yaml file from bash.

> Here is an example to illustrate the problem:
> 
> ```
> Sample secret set on GitHub Actions: `^!@#$%^&*pas$word123!^%!#`
> Reads from the workflow file "${{ secrets.MYSEC }}": returns `^!@#$%^&*pas!^%!#`
> ```

Bash script injection, your heredoc uses `<< EOF` (same applies to "") instead of `<< 'EOF'` this means bash expands `$`. I'm not shure why `$%^` survived.

```bash
          cat << EOF > "$HOME/.config/openstack/clouds.yaml"
          clouds:
            vex:
              auth:
                project_name: "${{ secrets.PROJECT_NAME }}"
                username: "${{ secrets.USERNAME }}"
                password: "${{ secrets.PASSWORD }}"
                auth_url: "https://auth.vexxhost.net/v3"
                user_domain_name: Default
                project_domain_name: Default
              region_name: "${{ secrets.REGION_NAME }}"
          EOF
```
yaml script injection: these secrets can inject new yaml keys.
Since yaml supports json syntax the following has no yaml injection bug.
Important are single quotes around the first EOF.
Your original code even allows to exit heredoc by injecting
```
"
EOF
```
into PROJECT_NAME. After tojson this is `"\"\nEOF"` and harmless.
```bash
          cat << 'EOF' > "$HOME/.config/openstack/clouds.yaml"
          clouds:
            vex:
              auth:
                project_name: ${{ tojson(secrets.PROJECT_NAME) }}
                username: ${{ tojson(secrets.USERNAME) }}
                password: ${{ tojson(secrets.PASSWORD) }}
                auth_url: "https://auth.vexxhost.net/v3"
                user_domain_name: Default
                project_domain_name: Default
              region_name: ${{ tojson(secrets.REGION_NAME) }}
          EOF
```

Only non UTF-8 bytes will cause encoding problems (Internally stored as UTF-16)

**Comment by askb** [comment, marked as answer]
@ChristopherHX thanks for the suggestion.  I've tried what was recommended using single quotes around 'EOF', but that never worked. 

The workaround was using base64 encode/decode and storing the entire config as a single secret instead of injecting the values runtime when setting up CLI utils config.

`
         mkdir -p "$HOME/.config/openstack"
         echo "${{ secrets.CLOUDS_YAML_B64 }}" | base64 --decode > "$HOME/.config/openstack/clouds.yaml"
`

There is some scope for improvement in terms of end-user experience, rather than having to escape the special chars ('$') or think about how many languages need to be interpreted in my steps.

**Reply by ChristopherHX** [reply]
If you generate yaml, json, hcl etc. from bash without proper escaping you will also see such issues without GitHub Actions beeing part of it. GitHub Actions literally replaces ${{ }} before invoking any shell. E.g. `shell: mv {0} config.json` won't expand $, because mv doesn't interpret any symbol of the generated script. (changing that would be a very big breaking change)

I assume your base64 config has not been generated by bash.
Yes encoding a proper config file in base64 works. I would guess your generated hcl file via bash was malformed too, but GitHub Actions has no builtin non json escaping rules.

If you are going to remove the non base64 secrets or the broken hcl generation code, you may see unmasked content from your hcl in logs. In this case GitHub Action doesn't know that fields of a decoded base64 secret needs masking, you can manually hint GitHub Actions by add-mask however then you have to parse hcl and escape again.

**Reply by askb** [reply]
Thank you for the reply Christopher.

There was nothing broken with the packer hcl files, work as expected locally, we were able to get this working only with using the base64 encode/decode through the job. The contents are all masked as expected which is nice.

https://github.com/opendaylight/releng-builder/actions/runs/5686455543/job/15413265605  

Well, not the desired end-user or developer experience not knowing that these situations aren't handled through the built-in functions/modules. It would be nice to have actions/workflow documentation updated for the benefit of future devs who stumble on the same issue in question.
